pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/steps/SecretStep.java

```diff
@@ -0,0 +1,152 @@
+            return "secret";
+        }
+
+        @Nonnull
+        @Override
+        public String getDisplayName() {
+            return "Provide an encrypted secret that will be decrypted in runtime";
+        }
+
+        @Override
+        public Step newInstance(@CheckForNull StaplerRequest req, @Nonnull JSONObject formData) throws FormException {
+            if (req != null && req.findAncestor(Snippetizer.class) != null) {
+                String text = formData.optString("text");
+                if (text != null) {
+                    try {
+                        formData.put("text", encrypt(text));
+                    } catch (Exception e) {
+                        throw new FormException("Could not encrypt the text: " + e.getMessage(), e, "text");
+                    }
+                }
+            }
+            return super.newInstance(req, formData);
+        }
+
+        @Restricted(NoExternalUse.class)
+        public static String decrypt(String secret) throws IOException, BadPaddingException, IllegalBlockSizeException {
+            byte[] bytes = Base64.decode(secret.toCharArray());
+            byte[] both = KEY.decrypt().doFinal(bytes);
+            byte[] prefix = new byte[SALT_LENGTH];
+            System.arraycopy(both, 0, prefix, 0, prefix.length);
+            byte[] suffix = new byte[SALT_LENGTH];
+            System.arraycopy(both, both.length - suffix.length, suffix, 0, suffix.length);
+            byte[] actual = new byte[both.length - (SALT_LENGTH*2) - 6];
+            System.arraycopy(both, SALT_LENGTH + 3, actual, 0, actual.length);
+            ArrayUtils.reverse(suffix);
+            if (!ArrayUtils.isEquals(prefix, suffix)) {
+                throw new StreamCorruptedException("Salt check differs! Are you using the original key?");
+            }
+            return new String(actual, "UTF-8");
+        }
+
+        @Restricted(NoExternalUse.class)
+        public static String encrypt(String plainText) throws IOException, BadPaddingException, IllegalBlockSizeException {
+            byte[] bytes = (":::" + plainText + ":::").getBytes("UTF-8");
+            byte[] salt = ConfidentialStore.get().randomBytes(SALT_LENGTH);
+            byte[] both = new byte[bytes.length + salt.length + salt.length];
+            System.arraycopy(salt, 0, both, 0, salt.length);
+            System.arraycopy(bytes, 0, both, salt.length, bytes.length);
+            ArrayUtils.reverse(salt);
+            System.arraycopy(salt, 0, both, salt.length + bytes.length, salt.length);
+
+            return new String(Base64.encode(KEY.encrypt().doFinal(both)));
+        }
+    }
+
+    public static final class StepExecutionImpl extends AbstractSynchronousNonBlockingStepExecution<String> {
+        private static final long serialVersionUID = 1L;
+
+        @Inject
+        private SecretStep step;
+
+        @Override
+        public String run() throws Exception {
+            return DescriptorImpl.decrypt(step.getText());
```

Comment:
Well the intention of this step as I understand it is to avoid having plain text secrets in your repo.
Whatever happens to it on the Jenkins master is of a lot lower concern. And afaiu the same this could be said about `withCredentials`.